PostgreSQL datasource

svenklemm opened the pull request:
#1542

I created this datasource by cloning the mysql datasource so the frontend is identical to the mysql frontend atm.

svenklemm committed:
add postgresql datasource

svenklemm committed:
add rest of files for postgres datasource

svenklemm committed:
fix timeseries query, remove unused code

svenklemm committed:
consistent naming, refactoring

svenklemm committed:
s/mysql/postgres/

svenklemm committed:
s/mysql/postgres/

svenklemm committed:
couple more tests

svenklemm committed:
tests for more datatypes

svenklemm committed:
fix macros for postgres

svenklemm committed:
add __timeSec macro

svenklemm committed:
add frontend for postgres datasource

svenklemm committed:
adjust documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=9209) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=9209) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=9209) it.</sub>

svenklemm committed:
fix formatting

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
add proper plugin description

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
merge editor changes from mysql

svenklemm committed:
port changes from mysql datasource

daniellee commented:
Thanks for this PR! This feature is on our roadmap to do so this is a great start. It is much appreciated you took the time to do this.

To be upfront about this - this PR will need a lot of work to get it to a mergeable state. This is not really anything to do with the code quality. The main reason for this is the code duplication between the MySQL and Posgres backends. We plan on adding more sql db data sources (ms sql server etc.) and would like to have an SQL data source abstraction that we can reuse.

The code in the MySQL plugin is definitely not perfect and we have already noticed a few things in your code that we would like to port over to the MySQL plugin. I'm going to test out the approach you took for `getTypedRowData` for example.

With PR's like this, there is always a risk it won't get merged as the refactoring step is hard to communicate via GitHub. This PR is quite functional already so I'd love to see it get merged. Here is how I think we should do:

We (the Grafana team) will give you feedback on the functionality (bugs or features that need to be fixed/changed/updated). Once the frontend and backend code is fully functional, we will push your branch to the grafana repo and work on extracting an abstraction for sql data sources.

Does this sound like it could work for you?

A summary of the feedback so far:
- The config page looks pretty good but there needs to be a way to disable ssl mode (we need to fix this in MySQL to allow ssl)
- When returning data in timeseries mode, only floats are accepted as a value. It should accept other types of numbers. 
- The start/placeholder query in the query editor is incorrect (it is the mysql placeholder query which does not work for postgres). For example, the postgres backend expects the time field to be called `time` rather than `time_sec` and there is no Postgres function called `UNIX_TIMESTAMP`.
- My first attempt at using macros didn't seem to work. I tried to use an epoch as the time field. I'll get back to you with more detailed feedback here. The  $__timeFilter macro with a timestamp seems to work.

More feedback to come!

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
set proper defaultQuery for postgres

svenklemm committed:
add time_sec to timeseries query
accept int for value for timeseries query

svenklemm committed:
revert allowing time_sec and handle int or float values as unix
timestamp for "time" column

svenklemm commented:
This works for me and i'd love to get this merged aswell.

I named the column "time" because it accepts datetimes aswell as unix timestamp. I could add time_sec aswell but I thought having a time column would be more generic since the code can handle unix timestamp and sql datetime.
I've replaced the placeholder query and integer values work as values too now. 
Which macro query did you not get to work?

svenklemm committed:
fix tslint error

svenklemm committed:
handle decimal values in timeseries query

svenklemm committed:
allow setting sslmode for postgres datasource

svenklemm committed:
use type switch for handling data types

jnicholls commented:
Thank you @svenklemm for working on this.  I'm probably not alone in saying that I'm going to be running with your work; the layering work @daniellee mentioned looks like it's in the 5.0 milestone and that's too far off to wait :)

svenklemm committed:
fix value for timeseries query

aplsek commented:
Hi, 

I am wondering how stable is this patch? 
I have built grafana with this plugin and am testing it. For a simple `SELECT  timestamp as "time_sec",  'charge' AS "metric" FROM ...` I am getting metric values displayed as `Y2hhcmdl`, the timestamp column is displayed correctly.

Thanks!

svenklemm commented:
Hi,

datatype handling needs some work as lib/pq doesnt handle alot of datatypes. Which version are you on? The latest version should handle char literals correctly but casting it to text explicitly should make it work too like 'charge'::text AS "metric" 
If you use datatypes supported directly by lib/pq it should work other datatypes might need some sql casting/extra code to handle them correctly.

aplsek commented:
I see, I am on `d39d30d` so I don't have the latest from `a73bcc5`.
I works now after the explicit cast.

I will be happy to test this more. Where should I post results/questions? Here?

svenklemm commented:
What did you have selected as query format? For time series it should work without cast and the code is already in your version. 
I guess here is the best place for results/questions unless the grafana team disagrees.

aplsek commented:
This was a Table with query format set as 'table'.

I am also testing a regular graph with the query format set as `timeseries`. There I can't get any data to display but that I think is a different problem, my query is :

svenklemm commented:
The time column needs to be named "time" returning unix timestamp is optional you can directly return sql timestamp
Example query:
```
select 'random' AS "metric",random() as "value",time from generate_series($__timeFrom(),$__timeTo(),$__interval::interval) AS "time"
```

aplsek commented:
I see, the MySQL instructions say to use `time_sec`. After the change it works. 
It prints a lot of debug messages but works.

The query that worked in the end.
```
SELECT
   extract(epoch from timestamp) as "time",
  'charge' AS "metric",
  charge AS "value"
FROM battery_status
WHERE $__timeFilter(timestamp) AND device_id = 'a6758'
ORDER BY timestamp asc
```

aplsek commented:
It seems to me there is an issue with timerange, when I select `Last 30 minutes` it shows only ~15 minutes worth of data. 
When I select `Last 1 hour` I get full hour of data.

svenklemm commented:
select "timestamp" as "time" should work too. I decided to name the column time cause it can handle sql timestamps directly aswell as unix timestamp.

aplsek commented:
Nice, `timestamp` works too - without the conversion, that is great! 

I do still have missing samples with my `Last 30 minutes`.

svenklemm commented:
Hmm that sounds weird the code doesnt remove any rows from the resultset but they could be out of sync with the display part. 
Next to the query format there is a button to show the sql after the macros have been evaluated. Can you run that in psql or pgadmin and see if it returns the expected values.

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
refactor timeseries queries to make them more flexible

svenklemm committed:
remove debug statement from inner loop in type conversion

svenklemm committed:
use plain for loop in getTypedRowData

svenklemm committed:
fix timeseries queries

svenklemm commented:
I've made the metric column optional, if its not specified the metric name is derived from the column name.
This allows creating multiple series from different columns in a single query. Having a column named "metric" is still supported but its no longer required.
Example query:
```
select 
random() as "random",
sin(time/100.0) as "sin",
cos(time/100.0) as "cos",
time from 
generate_series($__unixEpochFrom(),$__unixEpochTo(),10) AS "time"
```
This will create 3 series named "random","sin" and "cos"

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
adjust postgres datasource to tsdb refactoring

svenklemm committed:
adjust postgres datasource to frontend changes

svenklemm commented:
@daniellee I addressed all your feedback. 
There was an update of the postgres driver library couple days ago it would be useful to have the latest version as it makes more type information available in go. Should I open a separate PR for that or should I make it part of this one?

daniellee commented:
@svenklemm Fantastic! (Linking to that PR for reference: https://github.com/lib/pq/pull/647).

You can make it part of this PR. I will test this today and get back to you. Looks like you are getting close to being done.

svenklemm committed:
update lib/pq to latest version

svenklemm committed:
move type conversion to getTypedRowData

svenklemm committed:
handle address types cidr, inet and macaddr

svenklemm committed:
adjust response parser and docs for annotations

svenklemm committed:
convert unknown types to string

svenklemm committed:
add documentation for postgres datasource

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
add another example query with metric column

daniellee commented:
Just to let you know, I haven't forgotten you. I spent some testing this yesterday and functionality-wise this looks great (in fact I think we'll have to update the MySQL plugin to get it up to the same standard). I tried to break it using different types for datetimes and values but everything seemed to work :-)

I am out traveling (speaking at a conference) but am planning to start the merge work later in the week.

One thought so far. The placeholder query with generate_series, while great to see some data being graphed directly, I found it a little confusing when thinking about what the next step should be. It doesn't naturally lead me to writing a working query.

We should probably add some docs/macros for both Postgres and MySQL for grouping by time. I can start with MySQL as it is a bit trickier than it is for Postgres (which has date_trunc).

svenklemm committed:
set more helpful default query

svenklemm committed:
update help text in query editor

svenklemm commented:
I agree with you regarding the placeholder query, i replaced it with a query that doesnt use generate_series.

The documentation i committed in https://github.com/grafana/grafana/pull/9209/commits/61484606511f99be1daf0fa92bd4003f7e6a28ed
has example queries with dynamic group by time using $__interval even though those might not be optimal if your sample rate is infrequent. I guess having support for `min time interval` like some other datasources have would be beneficial.

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
handle NULL in value column of timeseries query

thomaspore commented:
All, 

Great work here!I'm excited for this to be part of master. I have built Grafana with this plugin and am testing it. I'm seeing a "Metric request error" when I "Save & Test" my PostgreSQL data source. Any one else run into this?
![grafana-pg_data_source_testing](https://user-images.githubusercontent.com/1752870/30983356-2d9995c2-a458-11e7-904f-2240f58d01fa.png)

svenklemm commented:
Can you set logging to debug and check if anything shows up in the error log. Maybe set sslmode explicitly
the default for sslmode is require so if your postgres server doesnt support ssl the test will fail unless you set sslmode to disable, but that should get you a different error message.

thomaspore commented:
Thank you for the quick response. Updated to SSL Mode to disabled to eliminate as a problem. Errors received:

EROR[09-28|15:06:26] Metric request error                     logger=context userId=1 orgId=1 uname=admin error="Could not find executor for data source type: postgres"
EROR[09-28|15:06:26] Request Completed                        logger=context userId=1 orgId=1 uname=admin method=POST path=/api/tsdb/query status=500 remote_addr=10.XXX.XXX.XXX time_ms=5 size=34 referer=http://10.XXX.XXX.YYY:3000/datasources/edit/3

svenklemm commented:
looks like your grafana doesnt know the postgres datasource did you do a proper rebuild?

thomaspore commented:
OK found the problem. I had run: 

```
curl -L https://github.com/grafana/grafana/pull/9209.patch > ./9209.patch
git apply --reject /root/9209.patch
```

I found main.go.rej with the following. 

```
diff a/pkg/cmd/grafana-server/main.go b/pkg/cmd/grafana-server/main.go  (rejected hunks)
@@ -24,8 +24,8 @@ import (
        _ "github.com/grafana/grafana/pkg/tsdb/influxdb"
        _ "github.com/grafana/grafana/pkg/tsdb/mqe"
        _ "github.com/grafana/grafana/pkg/tsdb/mysql"
-       _ "github.com/grafana/grafana/pkg/tsdb/postgres"
        _ "github.com/grafana/grafana/pkg/tsdb/opentsdb"
+       _ "github.com/grafana/grafana/pkg/tsdb/postgres"
        _ "github.com/grafana/grafana/pkg/tsdb/prometheus"
        _ "github.com/grafana/grafana/pkg/tsdb/testdata"
 )

```
To resolve I manually added main.go and rebuilt   
` _ "github.com/grafana/grafana/pkg/tsdb/postgres"`

All is well! I've got some basic queries running against my data. Thanks for the assistance!

svenklemm committed:
add __timeGroup macro

svenklemm committed:
add test for __timeGroup macro

svenklemm committed:
document __timeGroup and set proper default query for annotations

svenklemm commented:
@daniellee i've added a __timeGroup(column,interval) macro which can be used in a GROUP BY clause. You can either use it with a fixed interval `__timeGroup(column,'5m')` or pass $__interval as 2nd parameter `__timeGroup(column,$__interval)`

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

svenklemm committed:
fix typos in docs

svenklemm committed:
add postgres to list of datasources

svenklemm committed:
add postgres to builtInPlugins

svenklemm committed:
Merge remote-tracking branch 'upstream/master'

wazery commented:
Is there any built Docker images for this branch?

svenklemm commented:
@wazery i dont think there is a prebuilt docker image available
@daniellee do you need anything else from my side to merge this?

daniellee commented:
@svenklemm No, I have started abstracting out helper/base class and should be merging to master at the start of next week. So thanks for all your hard work!

@wazery there won't be any docker build until it is merged to master - which will be happening next week.

wazery commented:
@bergquist Since this PR is closed, isn't it going to be in the next release? Can I know the plan for the Postgres support. I tried the SqlDB plugin but it wasn't working, so this PR is my kind of my hope in using Grafana with Postgres data store.

torkelo commented:
It's not closed, but will be merged later coming week, hopefully beta release of 4.6 coming or next week

daniellee commented:
@svenklemm two followup questions for you:

- When you renamed the time_sec field for timeseries data to time, what were the reasons for that? I changed that back in #9475, am I messing anything up?
- Could you explain the __timeGroup macro a bit more. As a postgres noob, I'm not sure what the best approach is. Can you explain why you chose the extract(epoch from %s::interval) approach rather than date_trunc? And do you have an example query using it?

svenklemm commented:
I named it time because time is already used in a couple other places, e.g. annotation queries, influxdb time column is named time aswell and because it accepted datatypes other then unix timestamp but the name doesnt matter much and you can change it
date_trunc only allows truncation to specific precisions like second,minute,hour,day,week,... but you cannot do any values in between those like 5min or 15min and i guess with datetrunc you cant make it use $__interval to adjust the grouping depending on the selected time range in grafana.
Example Query:

select max(time) as time, avg(sin(extract(epoch from time)/1000)) from generate_series($__timeFrom(),$__timeTo(),'1s'::interval) as time(time)
group by $__timeGroup(time,$__interval) order by time asc

generate_series will create a timestamp for every second and the group by will aggregate them depending on the selected time range in grafana

jnicholls commented:
re: the use of `extract(epoch from <variable>)`, that is probably the most common way to pull out the full (epoch) seconds representation of any timestamp or interval variable.  However, another approach is `date_trunc('epoch', <variable>)` which is not as commonly seen (nor does it appear to be documented) but it works the same and I personally prefer it over the `extract` function.

torkelo commented on public/app/plugins/datasource/postgres/response_parser.ts:
```diff
@@ -0,0 +1,143 @@
+    for (let i = 0; i < table.columns.length; i++) {
+      if (table.columns[i].text === 'time') {
+        timeColumnIndex = i;
+      } else if (table.columns[i].text === 'title') {
```
the title column has been removed from annotations, they only have text field now

daniellee commented:
@jnicholls we want to keep it as similar as possible to the MySQL data source and the time column for my_sql is called time_sec. But after thinking about it some more, I think we will keep the name as time. Postgres is much more flexible than MySQL in this case.

The group by time function will probably be the source of a lot of issues. It is absolutely needed but just trying to understand it better in Postgres. Specialized timeseries databases like Graphite and InfluxDB do things like automatically aligning to the nearest hour, day etc. We will also need to add a macro for auto time grouping.

@svenklemm is this how you would use the macro?

```sql
SELECT
  min(date_time_col) AS time_sec,
  sum(value_double) as value
FROM yourtable
group by $__timeGroup(date_time_col, '1 hour')
order by $__timeGroup(date_time_col, '1 hour') ASC
```

daniellee commented:
Never mind saw something similar in the docs:

```sql
SELECT
  min(time_date_time) as time,
  min(value_double),
  'min' as metric
FROM test_data
WHERE $__timeFilter(time_date_time)
GROUP BY metric1, (extract(epoch from time_date_time)/extract(epoch from $__interval::interval))::int
ORDER BY time asc
```
Example with multiple columns:

```sql
SELECT
  min(time_date_time) as time,
  min(value_double) as min_value,
  max(value_double) as max_value
FROM test_data
WHERE $__timeFilter(time_date_time)
GROUP BY metric1, (extract(epoch from time_date_time)/extract(epoch from $__interval::interval))::int
ORDER BY time asc
```

I will add some help text for the $__timeGroup function to the metrics tab.

jnicholls commented:
@daniellee thanks!

re: the `$__timeGroup` macro, its usefulness is limited as it is currently implemented.  Your example as well as @svenklemm's example, it works as advertised because the projected fields (those in the select statement) are all aggregates.  However, the resulting value of the `$__timeGroup` is not useful as a `time` field, which is more than not the common case it will need to be used.

For example, if I want a time series of my error counts, summed into 15 minute buckets, I would expect to be able to do something like this:

```
SELECT
  $__timeGroup(ts, '15m') as time,
  sum(num_errors) as num_errors
FROM my_data
WHERE $__timeFilter(ts)
GROUP BY time
ORDER BY time;
```

However the value of the `$__timeGroup` macro as of now would only project the grouping value, and not a final timestamp value that can be used as plotting data on a time series chart.

To make it useful in this regard, it just needs to be updated to return the grouping value back into a timestamp by multiplying by the interval, and the net effect will be that the timestamp is rounded down to the nearest interval (e.g. for '15m' that would be 00:00, 00:15, 00:30, 00:45, 01:00, etc.)

This would be the updated code:

```
return fmt.Sprintf("to_timestamp((extract(epoch from \"%[1]s\")/extract(epoch from %[2]s::interval))::int * extract(epoch from %[2]s::interval))", args[0], args[1]), nil
```

Notice that we're effectively rounding down to the nearest interval by chomping off the remainder of the division, and then re-multiplying by the interval.  Finally, we take that epoch value and turn it into a timestamp (this is optional since we also support unix epoch secs).

I wrote an overloaded version of the `date_trunc` function that takes an interval instead of the expected values of 'minute', 'hour', 'day', etc. and does this same math.  Of course I'm using my `date_part` preference over the `extract` function, but it operates the same.

```
create or replace function date_trunc(i interval, ts timestamp) returns timestamp
as $$
    select to_timestamp(floor(date_part('epoch', ts) / date_part('epoch', i)) * date_part('epoch', i))::timestamp;
$$ language sql;

create or replace function date_trunc(i interval, ts timestamptz) returns timestamptz
as $$
    select to_timestamp(floor(date_part('epoch', ts) / date_part('epoch', i)) * date_part('epoch', i));
$$ language sql;
```

svenklemm commented:
__timeGroup is supposed to be used in the group by clause and you use min(time_column) or max(time_column) in the select to get the beginning or the end of the aggregation interval

```
SELECT
   min(ts) as time,
  sum(num_errors) as num_errors
FROM my_data
WHERE $__timeFilter(ts)
GROUP BY $__timeGroup(ts, '15m')
ORDER BY time;
```

jnicholls commented:
> __timeGroup is supposed to be used in the group by clause and you use min(time_column) or max(time_column) in the select to get the beginning or the end of the aggregation interval

Sure, I acknowledged that and it does work for that use case.  But extending it just a little further would make it equally as useful for that use case as well as be useful in the result set as the actual time value that is plotted in the time series data.

daniellee commented:
I think both functions would be useful. The graphite summarize functions have an align parameter: 

http://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.summarize

Think I can merge first and evolve this afterwards. It is already more feature rich than the MySQL data source.

jnicholls commented:
Yeah that's fair.

daniellee commented:
@svenklemm merged to master. Thanks for your amazing work on this PR!

svenklemm commented:
@daniellee Thanks, pleasure is all mine :)

aplsek commented:
I am testing the latest nightly build (4.6.0-9516pre1) and have problems with the Add Datasource form:

![screen shot 2017-10-10 at 12 15 27 pm](https://user-images.githubusercontent.com/888802/31406091-0040015e-adb5-11e7-976e-97d78e844d85.png)

ziXet commented:
Same. I pulled it from docker repo. Shouldn't be any field for password?

aplsek commented:
It works when I build grafana from source, so I guess/hope one of the next nightly builds should be fine.

daniellee commented:
@aplsek @ziXet Thanks for reporting this. Fixed now. Was missing an Angular decorator that messed up the config page after JavaScript minification (which is why it works when building from source).

So if you do a `docker pull grafana/grafana:master` before executing `docker run`, it should work now.

dturon commented:
Thanks all, I am very happy, now very simplify work when we can remove some layers :)
